src/services/pure-di.ts

```diff
@@ -0,0 +1,63 @@
+import axios, { AxiosStatic } from 'axios';
+import { HttpDopplerLegacyClient, DopplerLegacyClient } from './doppler-legacy-client';
+import { OnlineSessionManager, SessionManager } from './session-manager';
+
+interface AppConfiguration {
+  dopplerLegacyUrl: string;
+  dopplerLegacyKeepAliveMilliseconds: number;
+}
+
+export interface AppContainer {
+  axiosStatic: AxiosStatic;
+  appConfiguration: AppConfiguration;
+  dopplerLegacyClient: DopplerLegacyClient;
+  sessionManager: SessionManager;
+}
+
+export class AppCompositionRoot implements AppContainer {
+  private readonly instances: Partial<AppContainer>;
+
+  constructor(predefinedInstances: Partial<AppContainer> = {}) {
```

Comment:
`Partial<T>` is a type that has some (or none) of the properties of `T`, it allows us to accept a subset of _AppContainer_ properties, to override default _AppCompositionRoot_ services.

See more details in https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types